--- cryptlib.h	2020-07-07 10:35:32.235126000 +0200
+++ cryptlib.h.orig	2012-12-14 00:38:32.000000000 +0100
@@ -221,9 +221,6 @@
 	CRYPT_ALGO_RESERVED3,			/* Formerly MD4 */
 	CRYPT_ALGO_MD5,					/* MD5 */
 	CRYPT_ALGO_SHA1,				/* SHA/SHA1 */
-#ifdef KPYM_HACK
-	CRYPT_ALGO_SHA = CRYPT_ALGO_SHA1,	/* Older form */
-#endif
 	CRYPT_ALGO_RIPEMD160,			/* RIPE-MD 160 */
 	CRYPT_ALGO_SHA2,				/* SHA-256 */
 		CRYPT_ALGO_SHA256 = CRYPT_ALGO_SHA2,/* Alternate name */
@@ -232,9 +229,6 @@
 	/* MAC's */
 	CRYPT_ALGO_HMAC_MD5 = 300,		/* HMAC-MD5 */
 	CRYPT_ALGO_HMAC_SHA1,			/* HMAC-SHA */
-#ifdef KPYM_HACK
-	CRYPT_ALGO_HMAC_SHA = CRYPT_ALGO_HMAC_SHA1,	/* Older form */
-#endif
 	CRYPT_ALGO_HMAC_RIPEMD160,		/* HMAC-RIPEMD-160 */
 	CRYPT_ALGO_HMAC_SHA2,			/* HMAC-SHA2 */
 	CRYPT_ALGO_HMAC_SHAng,			/* HMAC-future-SHA-nextgen */
@@ -2326,20 +2320,6 @@
 						  C_IN HWND hWnd );
 
 #endif /* Win32 */
-#ifdef KPYM_HACK
-/****************************************************************************
-*																			*
-*							KpyM Hack Functions								*
-*																			*
-****************************************************************************/
-
-#define KEYEX_ALGO_LIST 0
-#define ENCR_ALGO_LIST 1
-#define MAC_ALGO_LIST 2
-#define KE_EMPTY_PASSWORD "hack za nqkoj e napisal prazna parola"
-C_RET cryptAlgoFromString( const char * str );
-C_RET cryptSetAlgoList( const int * list, int len, int type );
-#endif
 
 #ifdef __cplusplus
 }
