--- win32.c.orig	2012-07-25 07:40:26.000000000 +0200
+++ win32.c	2020-07-06 17:28:26.000000000 +0200
@@ -992,6 +992,20 @@
 		/* Remember that we've got the fixed info */
 		addedFixedItems = TRUE;
 		}
+#ifdef KPYM_HACK
+		{
+		LARGE_INTEGER performanceCount;
+
+		if( QueryPerformanceCounter( &performanceCount ) )
+			addRandomData( randomState, &performanceCount,
+						   sizeof( LARGE_INTEGER ) );
+		else
+			{
+			/* Millisecond accuracy at best... */
+			addRandomValue( randomState, GetTickCount() );
+			}
+		}
+#else
 
 	/* The performance of QPC varies depending on the architecture that it's 
 	   running on and on the OS, the MS documentation is vague about the 
@@ -1219,7 +1233,7 @@
 #endif /* 0 */
 		}
 #endif /* NO_ASM */
-
+#endif
 	/* Flush any remaining data through.  Quality = int( 33 1/3 % ) + any 
 	   additional quality from the TRNG if there's one present */
 	endRandomData( randomState, 34 + trngValue );
